File: backend/services/post-ktor-mongo-fdm/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/get_post/impl.kt
package me.jason5lee.post_ktor_mongo_fdm.get_post

import com.mongodb.client.model.Projections
import me.jason5lee.post_ktor_mongo_fdm.common.*
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps
import me.jason5lee.resukt.Result
import me.jason5lee.resukt.ext.asSuccess
import org.bson.types.ObjectId

class WorkflowImpl(private val deps: Deps) : Workflow() {
    override suspend fun run(id: PostId): Result<Post, Errors> {
        val objId = Db.toObjectIdOrNull(id.value) ?: return Result.failure(Errors.PostNotFound)
        val doc = Db.findById(
            db = deps.mongoDb,
            collection = Db.posts,
            id = objId,
            projection = null,
        ) ?: return Result.failure(Errors.PostNotFound)

        val title = Title(doc.get("title"))
        val content = run {
            val text = doc.getOptional<String>("text")?.let { TextPostContent(it) }
            val url = doc.getOptional<String>("url")?.let { UrlPostContent(it) }

            if (text != null && url == null) {
                PostContent.Text(text)
            } else if (text == null && url != null) {
                PostContent.Url(url)
            } else {
                throw Exception("Invalid query result from `${Db.posts}[_id = $objId]`, only exact one of `text` and `url` must present")
            }
        }
        val creator = doc.get<ObjectId>("creator")
        val creationTime = Time(doc.get("creationTime"))
        val lastModified = doc.getOptional<Long>("lastModified")?.let { Time(it) }

        val creatorDoc = Db.findById(
            db = deps.mongoDb,
            collection = Db.users,
            id = creator,
            projection = Projections.include("name"),
        )
            ?: throw Exception("Invalid query result from `${Db.posts}[_id = $objId].creator`, not found in `${Db.users}`")
        val creatorName = UserName(creatorDoc.get("name"))

        return Post(
            creator = Creator(
                id = UserId(Db.formatId(creator)),
                name = creatorName,
            ),
            creationTime = creationTime,
            lastModified = lastModified,
            title = title,
            content = content,
        ).asSuccess()
    }
}

File: backend/services/post-ts-koa-mongo/src/get-post/impl.ts
import { PostContent, PostId, UserId, TextPostContent, UrlPostContent, UserName, Title, Time } from "../common";
import { Deps } from "../common/utils";
import { Post, Workflow } from ".";
import * as db from "../common/utils/db";
import { ObjectId, WithId } from "mongodb";
import { errors } from "./api";
import * as runtypes from "runtypes";

export class WorkflowImpl implements Workflow {
    private static readonly expectedPost = runtypes.Record({
        creator: runtypes.InstanceOf(ObjectId),
        creationTime: runtypes.Number, // NodeJS will convert the small BSON Long to a number.
        lastModified: runtypes.Number.optional(),
        title: runtypes.String,
        text: runtypes.String.optional(),
        url: runtypes.String.optional(),
    });
    private static readonly userHasName = runtypes.Record({
        name: runtypes.String,
    });
    async run(id: PostId): Promise<Post> {
        const oid = db.tryParseId(id);
        if (oid === undefined) {
            throw errors.postNotFound();
        }
        const post: WithId<unknown> | null = await this.deps.mongoDb.collection(db.posts).findOne({ _id: oid });
        if (post === null) {
            throw this.errors.postNotFound();
        }
        db.validate(WorkflowImpl.expectedPost, db.posts, post);
        let content: PostContent;
        if (post.text !== undefined && post.url === undefined) {
            content = { type: "Text", content: post.text as TextPostContent };
        } else if (post.text === undefined && post.url !== undefined) {
            content = { type: "Url", content: post.url as UrlPostContent };
        } else {
            throw new Error(`Invalid value in collection \`${db.posts}\`, ID \`${JSON.stringify(id)}\`, exactly only one of \`text\` and \`url\` must exist`);
        }

        const creator: WithId<unknown> | null = await this.deps.mongoDb.collection(db.users).findOne({ _id: post.creator });
        if (creator === null) {
            throw new Error(`The creator of post ${id}, ${post.creator}, not found`);
        }
        db.validate(WorkflowImpl.userHasName, db.users, creator);
        return {
            creator: {
                id: db.formatId(creator._id) as UserId,
                name: creator.name as UserName,
            },
            title: post.title as Title,
            content,
            creationTime: { utc: post.creationTime } as Time,
            lastModified: post.lastModified !== undefined ? { utc: post.lastModified } as Time : undefined,
        };
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { }
}
